static-check/src/objects/symbol_table.rs

```diff
@@ -0,0 +1,492 @@
+        // Iterate over imported names.
+        for import in imports.iter() {
+            self.check_import_statement(import, import_parser)?;
+        }
+
+        Ok(())
+    }
+
+    ///
+    /// Checks that a given import statement imports an existing package.
+    ///
+    /// Additionally checks for duplicate imported names in the given vector of imports.
+    /// Types defined later in the program cannot have the same name.
+    ///
+    pub fn check_import_statement(
+        &mut self,
+        import: &ImportStatement,
+        import_parser: &ImportParser,
+    ) -> Result<(), SymbolTableError> {
+        // Check if the import name exists as core package.
+        let core_package = import_parser.get_core_package(&import.package);
+
+        // If the core package exists, then attempt to insert the import into the symbol table.
+        if let Some(package) = core_package {
+            return self.check_core_package(package);
+        }
+
+        // Attempt to insert the imported names into the symbol table.
+        self.check_package(import, import_parser)
+    }
+
+    ///
+    /// Inserts imported core package circuit names and types into the symbol table.
+    ///
+    /// Checks that the core package and all circuit names exist. Checks that imported circuit types
+    /// only contain known types.
+    ///
+    pub fn check_core_package(&mut self, package: &Package) -> Result<(), SymbolTableError> {
+        // Create list of imported core packages.
+        let list = CorePackageList::from_package_access(package.access.to_owned())?;
+
+        // Fetch core package symbols from `leo-core`.
+        let symbol_list = list.to_symbols()?;
+
+        // Insert name and type information for each core package symbol.
+        for (name, circuit) in symbol_list.symbols() {
+            // Store name of symbol.
+            self.insert_circuit_name(name.to_string(), ParameterType::from(circuit.clone()))?;
+
+            // Create new circuit type for symbol.
+            let circuit_type = CircuitType::new(&self, circuit.to_owned())?;
+
+            // Insert circuit type of symbol.
+            self.insert_circuit_type(circuit_type.identifier.clone(), circuit_type);
+        }
+
+        Ok(())
+    }
+
+    ///
+    /// Inserts one or more imported symbols for a given imported package.
+    ///
+    /// Checks that the package and all circuit and function names exist. Checks that imported circuit
+    /// and function types only contain known types.
+    ///
+    pub fn check_package(
+        &mut self,
+        import: &ImportStatement,
+        import_parser: &ImportParser,
+    ) -> Result<(), SymbolTableError> {
+        // Get imported symbols from statement.
+        let imported_symbols = ImportedSymbols::new(import);
+
+        // Import all symbols from an imported file for now.
+        // Keep track of which import files have already been checked.
+        let mut checked = Vec::new();
```

Comment:
if the number of symbols is expected to be higher than a few dozen, this should be a `HashSet` instead, in order for the look-ups in the loop below to not slow down in a linear fashion